### PR TITLE
Split the data into chunks and run those chunks in parallel

### PR DESCRIPTION
**Pull request by blob79**
I see a very modest overhead (~3%) for 1 CPU and a 3x improvement using 4 CPUs.

### CODE REVIEW

**Comment by blob79** [comment]
Added memmap::Mmap again in rev2.

**Comment by djc** [comment]
Are you testing on Linux or macOS? Someone suggested to me that memory mapping might not be as fast on macOS.

**Comment by djc** [comment]
(BTW, thanks for the continued improvements! I've called out your contribution in the README.)

**Comment by blob79** [comment]
Yes on macOS memory mapping is 30% slower.

**Comment by blob79** [comment]
I ran this change on a m6g.16xlarge EC2 instance (64 cpus, arm) with a 22 GB file on tempfs. It's doing ~12 GB/sec vs ~0.25GB/sec for master.

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 /target
 access_log
+*swp
```
Comment (djc) [inline .gitignore:3]: What is this for?

File: Cargo.toml [hunk 2]
```diff
@@ -10,12 +10,16 @@ categories = ["command-line-utilities", "text-processing"]
 edition = "2018"
 
 [dependencies]
+ahash = "0.3"
 anyhow = "1"
-err-derive = "0.2.4"
+err-derive = "0.2"
```
Comment (djc) [inline Cargo.toml:15]: Why did you change the err-derive version here?

File: src/bin/tf.rs [hunk 3]
```diff
@@ -1,21 +1,16 @@
-use std::fs::File;
-use std::io::Cursor;
-use std::str::FromStr;
-
 use anyhow::Error;
-use memmap::MmapOptions;
 use regex::Regex;
+use std::path::Path;
```
Comment (djc) [inline src/bin/tf.rs:3]: I prefer to keep the `std` imports separate (and these changes are not related to the topic of this commit/PR anyway.

File: src/lib.rs [hunk 4]
```diff
@@ -1,25 +1,45 @@
...
-        let ln = ln?;
-        if let Ok(key) = kf.key(&ln, &mut s) {
-            counter.add(key)
-        }
-    }
+) -> anyhow::Result<Vec<KeyCount>> {
+    let file = File::open(path)?;
+    let size = file.metadata()?.len();
+
+    let total = chunks::chunks(path, 1 << 26, size)
```
Comment (djc) [inline src/lib.rs:28]: What does this `1 << 26` constant mean?

File: src/chunks.rs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    files: Box<dyn Iterator<Item = anyhow::Result<Cursor<T>>> + Send>,
+    current: u64,
+    starts: std::vec::IntoIter<u64>,
+    chunk: usize,
+    size: u64,
+}
+
+pub fn chunks(path: &Path, chunk: usize, size: u64) -> Chunks<memmap::Mmap> {
+    let path = path.to_path_buf();
+    let it = (0..usize::MAX).map(move |_| {
```
Comment (djc) [inline src/chunks.rs:90]: I don't understand what the purpose of this loop is. In particular, why are we iterating over the `0..usize::MAX` range in the first place? Also why do we need this? If we've gotten the slice out of the memory map we can easily chunk that up.

File: src/chunks.rs [hunk 6]
```diff
@@ -0,0 +1,191 @@
+use anyhow::Context;
+use std::fs::File;
+use std::io::{BufRead, BufReader, Cursor, Seek, SeekFrom};
+use std::path::Path;
+
+fn split(chunk: u64, size: u64) -> Vec<u64> {
```
Comment (djc) [inline src/chunks.rs:6]: Please order the code (as in the order modules) in top-down order: top-level APIs come first, followed by things it depends on. `impl`s come after `struct` definitions.

File: src/chunks.rs [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+impl<T> Iterator for Chunk<T>
+where
+    T: Seek + BufRead,
+{
+    type Item = String;
+    fn next(&mut self) -> Option<String> {
+        if self.pos >= self.end {
+            return None;
+        }
+        match self.lines.next() {
```
Comment (djc) [inline src/chunks.rs:70]: You can use the try operator `?` to return `None` instead of matching explicitly (so this should translate to `let l = self.lines.next()?.ok()?;`.

File: src/chunks.rs [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+                self.pos += l.len() as u64 + 1;
+                Some(l)
+            }
+            _ => None,
+        }
+    }
+}
+
+pub struct Chunks<T> {
+    files: Box<dyn Iterator<Item = anyhow::Result<Cursor<T>>> + Send>,
```
Comment (djc) [inline src/chunks.rs:81]: Do we suddenly have multiple files? The top-level API suggests we don't. Maybe pick a different name for this?

File: src/chunks.rs [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+impl<T> Iterator for Chunks<T>
+where
+    T: AsRef<[u8]>,
+{
+    type Item = Chunk<BufReader<Cursor<T>>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(start) = self.starts.next() {
+            if let Some(Ok(f)) = self.files.next() {
```
Comment (djc) [inline src/chunks.rs:117]:
```suggestion
        let start = self.starts.next()?;
        let f = self.files.next()?.ok()?;
```

File: src/chunks.rs [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+impl<C> Chunk<C> {
+    pub fn new(
+        mut c: C,
+        chunk: usize,
+        mut current: u64,
+        start: u64,
+        size: u64,
+    ) -> anyhow::Result<(Self, u64)>
+    where
+        C: Seek + BufRead,
```
Comment (djc) [inline src/chunks.rs:27]: Why do we need to seek if we're memory mapping? We can just get the slice and access at will?

File: src/lib.rs [hunk 11]
```diff
@@ -1,25 +1,45 @@
-use std::io::BufRead;
+use rayon::iter::ParallelBridge;
+use rayon::prelude::*;
```
Comment (djc) [inline src/lib.rs:2]: I prefer not to have `*` imports (also please separate the `std` and `rayon` import blocks, `std` first).

File: .gitignore (hunk 1 above)
Comment (blob79) [reply .gitignore:3]: Removed again. It's a vim temp file.

File: src/bin/tf.rs (hunk 3 above)
Comment (blob79) [reply src/bin/tf.rs:3]: Done.

File: src/chunks.rs (hunk 5 above)
Comment (blob79) [reply src/chunks.rs:90]: We split the file in chunks. For this we open the file and seek to the start position. The iterator is a way to open unlimited number of files for the path. The asumption is that we have at most (in the order of) number of cpus files open at any given time as rayon is handing out work.

File: src/lib.rs (hunk 4 above)
Comment (blob79) [reply src/lib.rs:28]: `1<<26` is the same as `(2 as usize).pow(26)`.

File: src/chunks.rs (hunk 7 above)
Comment (blob79) [reply src/chunks.rs:70]: Done. That's much better.

File: src/chunks.rs (hunk 8 above)
Comment (blob79) [reply src/chunks.rs:81]: Done.

File: src/chunks.rs (hunk 9 above)
Comment (blob79) [reply src/chunks.rs:117]: Done. That's much better.

File: src/chunks.rs (hunk 10 above)
Comment (blob79) [reply src/chunks.rs:27]: Please have a look at the latest version, then it probably makes more sense. I want this to work for every type implements `T: Seek + BufRead` i.e. `BufReader` and `Cursor`. At the moment the actual code uses one the test the other.

File: src/lib.rs (hunk 11 above)
Comment (blob79) [reply src/lib.rs:2]: Done.

File: Cargo.toml (hunk 2 above)
Comment (blob79) [reply Cargo.toml:15]: I think that any 0.2 API should do. There's no need to be overly specific.

Comment (djc) [reply Cargo.toml:15]: I don't think this helps; please revert the change.

For versions whose major version is 0, Cargo treats the minor version as the major version (that is, minor version bumps are semver-incompatible). In turn, that means that both bug fixes and features go in micro version bumps, so it's never guaranteed that 0.2.0 will work if you've only tested with 0.2.4. As such, since 0.2.0 and 0.2.4 are semver-compatible, Cargo will never use anything below 0.2.4 for new projects if 0.2.4 once 0.2.4 has been released (except in the case that another crate in the same dependency graph specifies a precise dependency like `=0.2.2` -- which I think is exceedingly rare).

File: Cargo.toml [hunk 12]
```diff
@@ -10,12 +10,15 @@ categories = ["command-line-utilities", "text-processing"]
 edition = "2018"
 
 [dependencies]
+ahash = "0.3"
```
Comment (djc) [inline Cargo.toml:13]: Let's have these unrelated changes in a separate commit or PR.

File: src/key_finder.rs [hunk 13]
```diff
@@ -25,8 +25,8 @@ impl KeyFinder {
 
     pub fn key<'a>(&self, record: &'a str, s: &'a mut String) -> Result<&'a str, Error> {
         let (num, last, keep) = match &self.keys {
-            None => return Ok(record.into()),
-            Some((num, _, _)) if *num == 0 => return Ok(record.into()),
+            None => return Ok(record),
```
Comment (djc) [inline src/key_finder.rs:28]: Let's have these unrelated changes in a separate commit or PR.

File: src/lib.rs [hunk 14]
```diff
@@ -1,25 +1,45 @@
...
-    for ln in reader.lines() {
-        s.clear();
-        let ln = ln?;
-        if let Ok(key) = kf.key(&ln, &mut s) {
-            counter.add(key)
-        }
-    }
+) -> anyhow::Result<Vec<KeyCount>> {
+    let total = chunks(path, 1 << 26)?
+        .collect::<Vec<_>>()
```
Comment (djc) [inline src/lib.rs:27]: I don't think it makes sense here to first `collect()` the iterator into a `Vec` and then immediately turn it into an `Iterator` again using `into_iter()`?

File: src/lib.rs [hunk 15]
```diff
@@ -1,25 +1,45 @@
...
-    let mut s = String::new();
-    for ln in reader.lines() {
-        s.clear();
-        let ln = ln?;
-        if let Ok(key) = kf.key(&ln, &mut s) {
-            counter.add(key)
-        }
-    }
+) -> anyhow::Result<Vec<KeyCount>> {
+    let total = chunks(path, 1 << 26)?
```
Comment (djc) [inline src/lib.rs:26]: I still don't understand what the `1 << 26` value means, semantically. Please make it a named constant. It's also unclear why you use `1 << 26` rather than `1 << 25` or `1 << 27` (this could be explained in a comment). Wouldn't it make more sense to specify the number of chunks to aim for rather than a specific chunk size, and then use the number of (v)CPU's available by default?

File: src/chunks.rs (hunk 10 above)
Comment (djc) [reply src/chunks.rs:32]: There's no good reason for the memory not to pass a slice, so this isn't quite the explanation I was looking for.

File: src/counter.rs [hunk 16]
```diff
@@ -16,19 +16,23 @@ impl Counter {
         }
     }
 
-    pub fn add(&mut self, key: &str) {
+    pub fn counts(self) -> AHashMap<String, u64> {
```
Comment (djc) [inline src/counter.rs:19]: I think it'd be better to solve the merging of `Counter`s inside the `Counter` type, by adding `merge(&mut self, other: &Counter)` instead of this method.

File: src/chunks.rs (hunk 5 above)
Comment (djc) [reply src/chunks.rs:98]: Why don't we just open the `File` as part of the instantiation of a `Chunk`?

File: src/lib.rs [hunk 17]
```diff
@@ -1,25 +1,45 @@
...
+        .collect::<Vec<_>>()
+        .into_iter()
+        .par_bridge()
+        .map(|reader| {
+            let mut counter = Counter::new(0);
```
Comment (djc) [inline src/lib.rs:31]: Why did you use `0` as the `Counter::new()` argument here (and below in the `fold()` call)? It seems to me that it would be better to always use the cutoff.

File: Cargo.toml (hunk 2 above)
Comment (blob79) [reply Cargo.toml:15]: Ok.

File: src/chunks.rs (hunk 5 above)
Comment (blob79) [reply src/chunks.rs:98]: You need a path to open a file. In the tests I want to run in mem without file io, so I need some why to avoid refering to paths. There's probably a better way to do this.

File: src/chunks.rs (hunk 6 above)
Comment (blob79) [reply src/chunks.rs]: Done.

File: Cargo.toml (hunk 12 above)
Comment (blob79) [reply Cargo.toml:13]: Done.

File: src/key_finder.rs (hunk 13 above)
Comment (blob79) [reply src/key_finder.rs:28]: Done.

File: src/lib.rs (hunk 14 above)
Comment (blob79) [reply src/lib.rs:27]: It doesn't. I couldn't get this to work without collect.

File: src/lib.rs (hunk 15 above)
Comment (blob79) [reply src/lib.rs:26]: With rayon's work stealing implementation you want to have more chunks (10x) than CPUs to make the CPU utilization more even. I'll change the implementation to refect this.

File: src/counter.rs (hunk 16 above)
Comment (blob79) [reply src/counter.rs:19]: Done

File: src/lib.rs (hunk 17 above)
Comment (blob79) [reply src/lib.rs:31]: When you map over the lines you don't need to calculate top at all. It's only relevant to update counts.

Only once you fold it makes sense to keep track of top - even then some top computation are discarded.

File: src/chunks.rs (hunk 10 above)
Comment (blob79) [reply src/chunks.rs:32]: As I've mentioned the implemenation works for files and in memory bytes.

File: src/lib.rs (hunk 17 above)
Comment (djc) [reply src/lib.rs:22]: In that case, what if we make `num` an `Option`? In that case we can also avoid the unnecessary `top` updates.

Comment (blob79) [reply src/lib.rs:22]: Ok make sense to use Option to make this explicit.

**Comment by djc** [comment]
Thanks for the improvements!

**Comment by djc** [comment]
I've made a bunch of further changes. These reduce the amount of code in the aggregate, fix a bug and even increase the performance by a few more %. Your review is welcome:

https://github.com/djc/topfew-rs/compare/e5e9287..28c39ca

**Comment by blob79** [comment]
Looks good to me. Some good improvements in their.
  
You really don't like bitshifting. For your entertainment: I'd say `1<<10`, `1<<20`, `1<<30` are more readable as they are kb, Mb, Gb. Knowing this, you know the order of magnitude of a value like `1<<24`: `1<<4 * 1 Mb`. That's better than `16777216` or `16 * 1024 * 1024`. Maybe `16 * 1<<20` has the best readability - short and you don't not have to do `1<<4` in your head.

**Comment by djc** [comment]
I do find `16 * 1024 * 1024` much easier to understand than `1 << 26`, yes, and I strongly suspect this will be true for most developers. But, to validate this I've just posted a Twitter poll: https://twitter.com/djco/status/1267719452025671680.
